ramips: add support for TP-Link RE210 v1

achterin opened the pull request:
Hardware
--------

SoC:   MediaTek MT7620A
RAM:   64MB
FLASH: 8MB SPI
WLAN:  2G: MediaTek MT7620A
       5G: MediaTek MT7610EN
ETH:   1x 10/100/1000M
LED:   RSSI (orange/green)
       WiFi 2G (green)
       WiFi 5G (green)
       Power (green)
       System (red / green)
BTN:   Power
       Reset
       LED
       WPS

Serial
------

P1 - Tx
P2 - Rx
P3 - GND
P4 - VCC

Pin 4 is the one closest to the LAN port.

Installation
------------

Web interface:
It is possible to upgrade to OpenWrt via the web interface. However, the
OEM firmware upgrade file is required and a tool to fix the MD5 sum of
the header. This procedure overwrites U-Boot and there is not failsafe /
recovery mode present! To prepare an image, you need to take the header
and U-Boot (i.e. 0x200 + 0x20000 bytes) from an OEM firmware file and
attach the factory image to it. Then fix the header MD5Sum1.

Serial/TFTP:
You can use initramfs for booting via RAM or flash the image directly.

Additional Notes:
If the web interface upgrade fails, you have to open your device and
attach serial console. Since the web upgrade overwrites the boot loader,
you might also brick your device.

In order to flash back to stock, the first header and U-Boot needs to be
stripped from the original firmware.

Signed-off-by: Christoph Krapp <achterin@googlemail.com>

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts:
```diff
@@ -0,0 +1,159 @@
+	keys {
+		compatible = "gpio-keys";
+
+		led-power {
```
Use underscore (led_power). For the the label property I'd consider "LED power", though this is optional.

achterin replied:
done

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts:
```diff
@@ -0,0 +1,159 @@
+};
+
+&ethernet {
+	mtd-mac-address = <&config 0x158>;
```
Do you find the same value in u-boot partition? (Typically 0x1fc00 ?).
config shouldn't be used, as this is typically created by stock FW and not present from the beginning.

achterin replied:
you're right, the base mac can be found at this address. i thought i'd changed that before and haven't found anything useful there....guess i mistyped something. thanks for the hint.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts:
```diff
@@ -0,0 +1,159 @@
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&radio 0x0>;
```
I suppose there is nothing useful in 0x4 and 0x8004 of radio partition?

Please provide a MAC overview in your commit message (similar to https://openwrt.org/docs/guide-developer/mac.address#merge_your_data)

achterin replied:
nope, nothing remotely close to the device mac there.

andyboeh commented:
There seem to be a lot of similarities between the RE200v1 and the RE210v1. Should we consider creating a shared .dtsi for both devices?

achterin commented:
@andyboeh as your pr got merged, i will take care of the unification.
i just experienced some trouble with the gigabit part of the ethernet that i need to sort out first.

adrianschmutzler commented:
@achterin @andyboeh It would be good if one of you tested my patch for moving mktplinkfw, too:
https://patchwork.ozlabs.org/patch/1216790/

achterin commented:
@adrianschmutzler will do

blocktrron commented on target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts:
```diff
@@ -0,0 +1,159 @@
+};
+
+&gpio2 {
+	status = "okay";
```
This GPIO controller seems to be unused

andyboeh commented:
@achterin Did you check whether sysupgrade works? I just noticed that I cannot do a sysupgrade to the latest snapshot on the RE200v1, because a call to `append-metadata` is missing. See https://github.com/openwrt/openwrt/pull/2672.

achterin commented:
@andyboeh yeah, i've run into that issue as well. will hopefully fix that as well as the gigabit link issue on monday.

cnbr commented:
@achterin  compiled, installed it on a re210 via the stock web interface, works.
I use the serial console for the first configuration and tests.
Connection via the 2.4 GHz interface works, with the 5 GHz interface there is no connection. However, an `iwlist scan `shows networks at 5 GHz.
LAN interface does not get a connection.
LEDs work, only the RSSI Led has a green and red LED, it lights up orange when red and green are on at the same time.
Thank you for your work, the re210 will become a usable device

blocktrron commented:
@cnbr Did you built from achterins tree or the commit picked on the current master? There was an issue with mt76 which allowed no connection to a WiFi network with patched EEPROM yesterday which has been fixed in the meantime.

achterin commented:
@cnbr thanks for testing my pr. lan does work if you limit your ethernet speed to 100mbit/s, gigabit seems broken. 5ghz wifi worked fine for me, confirmed via phone and notebook.

blocktrron commented:
@achterin From the FCC filing, the device uses an external AR8035 PHY. When testing, plese yop this patch to the ramips target, as otherwise the RGMII delays in the PHY might be incorrectl set: 

https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/patches-4.14/0038-at803x-disable-delays.patch

achterin commented:
@blocktrron as i'm pretty new to kernel patching in openwrt, what would be your preferred way here: applying this patch in the ramips target as well (thus keeping a second copy of the patch) or removing it from ath79 and move it to the generic patchset as some at803x-patches already reside there (734-736)

adrianschmutzler commented:
I'd consider creating a shared DTSI for this one and RE200 v1.

achterin commented:
@adrianschmutzler tested your patchset but have to report that with it, no sysupgrade & factory images will be created for my device.

@blocktrron i've ported the patch to rampis but the connectivity issue remains. i also tried a bunch of dts changes but that got me nowhere. i've tested the 100mbit/s-limited link again and it doesnt work anymore as well. the tplink re350k-v1 uses the same atheros chip but i dont know how to port the necessary changes to this soc. any idea?

cnbr commented:
@blocktrron I used the current master,  and made the changes manually. repeated it today. 5GHz works now. 
@achterin LAN does not work, check it with e 100MBit Port, ifconfig shows always "RX bytes:0 (0.0 B)", TX Byte are counted up

adrianschmutzler commented:
> tested your patchset but have to report that with it, no sysupgrade & factory images will be created for my device.

I forgot to move rootfs_align as well, that's only been present in ath79/mpc85xx. Fixed it now and was able to build sysupgrade/factory images for RE200v1. Can you retest with my updated staging tree?
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

(Note that you will have to change "mktplinkfw" to "tplink-v1-image" in mt7620.mk)

achterin commented:
@cnbr yeah, i've tested everything once again yesterday and couldn't figure out how to solve the link issue. i was pretty sure lan was working on 100mbit/s while developing during 36c3 but apparently my memory tricked me here.

@adrianschmutzler picked your changes and now both images are generated. will test the images later today and report if they do work as well.

blocktrron commented:
@achterin The patch fixes an issue with the internal RGMII delays of the Atheros PHYs. As this potentially breaks existing boards, I've only ported it to ath79 ATM. However, upstream has this corrected starting with kernel 5.1, so it will come in the future.

In short - the `phy-mode` property on your MAC controls whether or not the PHY activates internal delays. An (IMHO) pretty good explanation can be found here: https://web.archive.org/web/20190930022038/https://ethernetfmc.com/rgmii-interface-timing-considerations/

For the MT7620 you need something like this (Stolen from the Buffalo WHR-1166D):

```
&ethernet {
	pinctrl-names = "default";
	pinctrl-0 = <&rgmii1_pins &mdio_pins>;

	mtd-mac-address = <&uboot 0x1fc00>;

	port@5 {
		status = "okay";
		phy-handle = <&phyPHYID>;
		phy-mode = "RGMIIMODE";
	};

	mdio-bus {
		status = "okay";

		phyPHYID: ethernet-phy@PHYID {
			reg = <PHYID>;
		};
	};
};
```

You'll need to determine the ID of the PHY on the MDIO bus. This is either referenced somewhere in the GPL code or you can find the ID by determining which bootstrap options are set on the PHY. The datasheet can be found via Google.

The phy-mode is either referenced somewhere in the GPL code or you can do trial and error (It's only 4 possible values): https://elixir.bootlin.com/linux/v4.18.20/source/Documentation/devicetree/bindings/net/ethernet.txt

achterin commented:
@adrianschmutzler the generated images do work with your patch. you can add this to the tested device list ;)

@blocktrron awesome, thank you for your detailed explanation. this will not only help me with this device but with another one as well! i'll try to find the proper values within the next days. the dts changes i've tried looked something like this but the values were obviously wrong.

adrianschmutzler commented:
@achterin Thanks for testing. I've added your Tested-by and merged my patch, thus you would need to update this PR to use tplink-v1-image now.

andyboeh commented:
@achterin What TX power settings are available for you on the 5GHz band? On my RE200, the 2.4GHz WiFi has up to 20dBm available, while `iw list` only report 12dBm on the 5GHz band (it works, but the range is severely limited).

achterin commented:
@andyboeh 2.4GHz on my device also supports up to 20dBm while 5GHz supports up to 15dBm.

andyboeh commented:
@achterin Yeah, I was wrong: the range is quite normal, seems like it doesn't support more than 12dBm on the 5GHz band.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -974,6 +974,25 @@ define Device/tplink_re200-v1
+  TPLINK_HEADER_VERSION := 1
+  TPLINK_FLASHLAYOUT := 8Mmtk
+endef
+TARGET_DEVICES += tplink_re210-v1
```
Hi, be aware of https://patchwork.ozlabs.org/patch/1231065/ which would reduce the entries in this block:

```
define Device/tplink_re210-v1
  $(Device/tplink-v1)
  SOC := mt7620a
  DEVICE_MODEL := RE210
  DEVICE_VARIANT := v1
  DEVICE_PACKAGES := kmod-mt76x0e
  IMAGE_SIZE := 7936k
  TPLINK_HWID := 0x02100001
  TPLINK_FLASHLAYOUT := 8Mmtk
endef
TARGET_DEVICES += tplink_re210-v1
```

adrianschmutzler replied:
Has been merged in the meantime.

achterin replied:
Seen, thx. Will continue to work on this PR in about 2 weeks.

Yetangitu commented:
The RE210 only has a single external PHY on port 4 at mdio-bus:04:

```
[    5.113475] mtk_soc_eth 10100000.ethernet: connected port 4 to PHY at mdio-bus:04 [uid=004dd072, driver=Atheros 8035 ethernet]
```

It can be enabled by setting port 4 to `gmac` and removing `rgmii1` and `mdio` from `gpio`:

```
&state_default {
	gpio {
		ralink,group = "i2c", "uartf", "wled", "spi refclk";
		ralink,function = "gpio";
	};
};

&gsw {
        mediatek,port4 = "gmac";
};

&ethernet {
	status = "okay";
	pinctrl-names = "default";
	pinctrl-0 = <&rgmii1_pins &rgmii2_pins &mdio_pins>;

	port@4 {
		status = "okay";
		phy-handle = <&phy4>;
		phy-mode = "rgmii";
	};

	mdio-bus {
		status = "okay";

		phy4: ethernet-phy@4 {
			reg = <4>;
			phy-mode = "rgmii";
		};
	};
};
```

A working (for this specific problem... other problems remain) `.dts` can be found here:

https://github.com/Yetangitu/openwrt/blob/tplink-re210/target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts

Remaining problems can be gleaned from the kernel log:

```
[    5.311588] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    5.326998] rt2880-pinmux pinctrl: pin 24 is not set to gpio mux
[    5.339055] rt2880-pinmux pinctrl: request() failed for pin 24
[    5.350682] rt2880-pinmux pinctrl: pin-24 (pio:24) status -22
[    5.362135] gpio-keys keys: unable to claim gpio 24, err=-22
[    5.373445] gpio-keys: probe of keys failed with error -22
[    5.386488] rt2880-pinmux pinctrl: pin 27 is not set to gpio mux
[    5.398545] rt2880-pinmux pinctrl: request() failed for pin 27
[    5.410171] rt2880-pinmux pinctrl: pin-27 (pio:27) status -22
[    5.421645] leds-gpio: probe of leds failed with error -22
[    5.433595] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
```
pin 24 is `LED_POWER_SWITCH_ON_OFF` in the original GPL source distribution, pin 27 is `GPIO_POWER_SYSTEM_LED`, both pins from gpio1.

Yetangitu commented:
FYI, I flashed a build with the changes I mentioned above to the device and have been using this as an access point for a while, as far as I can tell the essential bits work as they should. As to whether the switches and LEDs work as intended I do not know, this is where some more testing might be needed.

Yetangitu commented:
![2020-02-17_21-19](https://user-images.githubusercontent.com/2754838/74684106-4e2cef00-51cb-11ea-855c-7bb742c4352d.png)

cnbr commented:
@Yetangitu, i build your `.dts` with the current master. LAN (testet 100/1000 MBit) and and both WiFi are working.  As described, the LEDs and switches unfortunately do not work properly.

Unfortunately, I don't know how to adjust the DTS files so that the problems with the rt2880-pinmux are solved.

achterin commented:
> It can be enabled by setting port 4 to `gmac` and removing `rgmii1` and `mdio` from `gpio`:

@Yetangitu thats not entirely correct but nevertheless thank you so much for your help. rgmii1 is not needed for lan and mdio is not needed for anything else but lan. so i just removed mdio from the gpio list, used the rest of your snippet and removed &rgmii1_pins from pinctrl-0, done. i will push an updated commit shortly after testing everything again to make sure it works.

achterin commented:
@adrianschmutzler can you please review this again? wifi, lan, buttons and leds do work now and i've merged re200v1 & re210v1 as requested. i hope i've catched all relevant changes to the target since december. thanks in advance

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re200-v1.dts:
```diff
@@ -1,7 +1,7 @@
+#include "mt7620a_tplink_re2x0-v1.dtsi"
 
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/input/input.h>
```
Those two lines can be removed, as they are in the dtsi as well.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts:
```diff
@@ -0,0 +1,91 @@
+
+&keys {
+	led_power {
+		label = "LED power";
```
So, this is a key that's used to turn the LEDs on or off?

achterin replied:
correct. with vendor firmware this button is used to enable/disable the leds.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts:
```diff
@@ -0,0 +1,91 @@
+
+&state_default {
+	gpio {
+		ralink,group = "i2c", "uartf", "wled", "rgmii1", "spi refclk";
```
Why "spi refclk"?

achterin replied:
good catch, not needed. removed it.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re2x0-v1.dtsi:
```diff
@@ -0,0 +1,98 @@
+
+&gpio3 {
+	status = "okay";
+};
```
Since they are determined by the choice of LEDs, I'd keep those (gpio1, gpio3) to the DTS files.

achterin replied:
done

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts:
```diff
@@ -0,0 +1,91 @@
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
+		label-mac-device = &ethernet;
```
You could consider to the label-mac-device to the dtsi:

```
	aliases {
		label-mac-device = &ethernet;
	}
```

achterin replied:
done

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts:
```diff
@@ -0,0 +1,91 @@
+		rssi2 {
+			label = "re210-v1:orange:rssi";
+			gpios = <&gpio1 2 GPIO_ACTIVE_LOW>;
+		};
```
Optional:
Consider to rename those to rssi_low and rssi_high or something else being more close to how they are used.
I'd typically expect the orange LED to be "low", so having this called rssi2 is somewhat unexpected to me.
But since they are not used at the moment anyway, that's purely cosmetic.

achterin replied:
done. changed them to the suggested names and also used "red" for rssi_low, as thats the correct color of the led.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts:
```diff
@@ -0,0 +1,91 @@
+
+	port@4 {
+		status = "okay";
+		phy-handle = <&phy4>;
```
Add empty line after status.

achterin replied:
done

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts:
```diff
@@ -0,0 +1,91 @@
+};
+
+&ethernet {
+	status = "okay";
```
Add empty line after status.

achterin replied:
done, removed

adrianschmutzler commented:
@achterin I've added this to my staging tree and applied the cosmetic changes myself. Please have a look at remarks nevertheless. If you want to "improve" the rssi LED names, just give feedback and I will apply it directly in the staging tree. It would be nice if you could also test the final version:

https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/staging

@andyboeh Would you be so kind to build my staging tree for RE200v1 and test it, so we have not destroyed anything during DTSI introduction?

adrianschmutzler commented:
@blocktrron Will we still need to port https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/patches-4.14/0038-at803x-disable-delays.patch ?

andyboeh commented:
> @andyboeh Would you be so kind to build my staging tree for RE200v1 and test it, so we have not destroyed anything during DTSI introduction?

I will try my best, but I won't have time for this before the weekend.

andyboeh commented:
@adrianschmutzler I just did a sysupgrade on my RE200v1 using your staging tree, everything works fine. The first boot took ages, but configuration files were preserved and everything seems to work as intended.

andyboeh commented:
@achterin @adrianschmutzler I also tested the build for the RE210 and I could successfully flash a specially prepared image via the OEM web interface. So far, it seems to work fine.

Yetangitu commented:
FYI, there is one problem I see with this port on my device:

```
[   29.988083] jffs2_scan_eraseblock(): End of filesystem marker found at 0x0
[   30.021092] jffs2_build_filesystem(): unlocking the mtd device... 
[   30.021167] done.
[   30.037366] jffs2_build_filesystem(): erasing all blocks after the end mark 
[   30.037389] jffs2: Erase at 0x00418000 failed immediately: errno -22
```
Reducing the size of the partition does not make a difference, it still fails to erase the last block:
```
[   38.965326] jffs2_scan_eraseblock(): End of filesystem marker found at 0x0
[   38.999807] jffs2_build_filesystem(): unlocking the mtd device... 
[   38.999879] done.
[   39.016087] jffs2_build_filesystem(): erasing all blocks after the end mark 
[   39.016110] jffs2: Erase at 0x00358000 failed immediately: errno -22
```
Most settings are saved but the root password is reset when the device is unplugged. Is anyone else seeing this behaviour?

achterin commented:
@adrianschmutzler sorry for the delay and thanks for the review. changed all suggested parts and retested the device. i also retested it after reverting commit 15a0701cdde8eeae2a54880b813cdb8cdc09a384 and can confirm that it solves the problem reportet by @Yetangitu.
furthermore i dont think we need https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/patches-4.14/0038-at803x-disable-delays.patch as lan works fine on my device without it but maybe @blocktrron can elaborate on why this patch might be needed.

achterin committed:
ramips: add support for TP-Link RE210 v1

Hardware
--------

SoC:   MediaTek MT7620A
RAM:   64MB
FLASH: 8MB SPI
WLAN:  2G: MediaTek MT7620A
       5G: MediaTek MT7610EN
ETH:   1x 10/100/1000M (Atheros AR8035)
LED:   RSSI (orange/green)
       WiFi 2G (green)
       WiFi 5G (green)
       Power (green)
       System (red / green)
BTN:   Power
       Reset
       LED
       WPS

Serial
------

P1 - Tx
P2 - Rx
P3 - GND
P4 - VCC

Pin 4 is the one closest to the LAN port.

MAC overview
------------

WAN *:4c uboot 0x1fc00
2.4 *:4c uboot 0x1fc00
5   *:4e uboot 0x1fc00 +2

Installation
------------

Web interface:
It is possible to upgrade to OpenWrt via the web interface. However, the
OEM firmware upgrade file is required and a tool to fix the MD5 sum of
the header. This procedure overwrites U-Boot and there is not failsafe /
recovery mode present! To prepare an image, you need to take the header
and U-Boot (i.e. 0x200 + 0x20000 bytes) from an OEM firmware file and
attach the factory image to it. Then fix the header MD5Sum1.

Serial/TFTP:
You can use initramfs for booting via RAM or flash the image directly.

Additional Notes:
If the web interface upgrade fails, you have to open your device and
attach serial console. Since the web upgrade overwrites the boot loader,
you might also brick your device.

In order to flash back to stock, the first header and U-Boot needs to be
stripped from the original firmware.

Signed-off-by: Christoph Krapp <achterin@googlemail.com>

blocktrron commented:
@achterin This patch is not needed (strictly speaking), but with it you can verify whether or node the correct RGMII delay setting is used for the device.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_tplink_re210-v1.dts:
```diff
@@ -0,0 +1,95 @@
+		};
+
+		rssi_high {
+			label = "re210-v1:green:rssi_high";
```
I'd prefer hyphens here (rssi-high) as this matches the style of the model name. For the node name, please keep the underscore. Same for rssi_low

adrianschmutzler commented:
Looks good to me except for the single comment above. Most probably won't have a chance to merge that before the weekend.

adrianschmutzler commented:
Thanks, merged (with the label adjustment).
